File: Roadmap/09 - HERENCIA/python/rikmij.py
class Animal:
    patas = 4
    cola = True
    
    def talk(self):
        pass

class Perro(Animal):
    def talk(self):
        return "<Guau>"

class Gato(Animal):
    def talk(self):
        return "<Miau>"

canino = Perro()
print(canino.talk())
print(canino.patas)

michi = Gato()
print(michi.talk())
print(michi.cola)


print('\n', '~'*7, "EJERCICIO EXTRA", '~'*7)

class Employee:
    def __init__(self, name, id, employment, employee):
        self.name = name
        self.id = id
        self.employment = employment
        self.employee = employee
    
    def meet_employee(self):
        return f"{self.name}: id - {self.id}"
    
    def get_employees(self):
        print(f"\n-> {self.name} está a cargo de:")
        for employ in self.employee:
            print(f"{employ.name} ({employ.employment}): id - {employ.id}")

class Programmer(Employee):
    def __init__(self, name, id, employment, language, employee):
        super().__init__(name, id, employment, employee)
        self.language = language
    
    def work(self):
        return f"{self.name}: Estoy trabajando en {self.language}"

class ProjectManager(Employee):
    def __init__(self, name, id, employment, employee, project):
        super().__init__(name, id, employment, employee)
        self.project = project
    
    def supervise(self):
        return f"{self.name}: Estoy \"supervisando\" {self.project}. ¿Cómo vais muchachos?"
    
class Manager(Employee):
    def __init__(self, name, id, employment, employee, organization):
        super().__init__(name, id, employment, employee)
        self.organization = organization
    
    def organize(self):
        return f"{self.name}: Estoy organizando {self.organization}"
    
victor = Programmer("Víctor", 11, "Programador", "Python", [])
joel = Programmer("Joel", 12, "Programador", "Kotlin", [])
print(victor.work())
print(joel.meet_employee())

programmer_list = []
programmer_list.append(victor)
programmer_list.append(joel)

giovanni = ProjectManager("Giovanni", 13, "Project Manager", programmer_list, "COBOL Multiplatform")
giovanni.get_employees()
print(giovanni.supervise())

boss_list = []
for emp in programmer_list:
    boss_list.append(emp)
boss_list.append(giovanni)

guido = Manager("Guido Van Rossum", 14, "Manager", boss_list, "Python")
guido.get_employees()
print(guido.organize())

File: Roadmap/09 - HERENCIA/kotlin/rikmij.kt
open class Animal {
    val patas = 4
    val cola = true
    open fun talk() {
    }
}

class Perro : Animal() {
    override fun talk() {
        println("Guau")
    }
}

class Gato : Animal() {
    override fun talk() {
        println("Miau")
    }
}

open class Employee(
    val name: String, private val identif: Int, private val employment: String,
    private val employees: MutableList<Employee>
) {
    fun meetEmployee() {
        println("$name ($employment): id - $identif")
    }

    fun showEmployees() {
        println("$name está a cargo de:")
        for (employee in employees) {
            println("${employee.name} (${employee.employment}), identificador: ${employee.identif}")
        }
    }
}

class Programmer(
    name: String, identif: Int, employment: String, employees: MutableList<Employee>,
    private val language: String
) : Employee(name, identif, employment, employees) {

    fun work() {
        println("-> $name: Estoy programando en $language")
    }
}

class ProjectManager(
    name: String, identif: Int, employment: String, employees: MutableList<Employee>,
    private val project: String
) : Employee(name, identif, employment, employees) {
    fun supervise() {
        println("\n-> $name: Estoy \"supervisando\" $project. ¿Cómo vais muchachos?")
    }
}

class Manager(
    name: String, identif: Int, employment: String, employees: MutableList<Employee>,
    private val organization: String
) : Employee(name, identif, employment, employees) {
    fun organize() {
        println("\n-> $name: Estoy organizando en $organization")
    }
}


fun main() {
    val canino = Perro()
    canino.talk()
    println(canino.patas)

    val michi = Gato()
    michi.talk()
    println(michi.cola)

    println("${"\n" + "~".repeat(7)} EJERCICIO EXTRA ${"~".repeat(7)}")

    val employeeList = mutableListOf<Employee>()
    val victor = Programmer("Víctor", 11, "Programador", mutableListOf(), "Python")
    victor.work()
    val joel = Programmer("Joel", 12, "Programador", mutableListOf(), "Java")
    joel.meetEmployee()

    employeeList.add(victor)
    employeeList.add(joel)

    val giovanni = ProjectManager(
        "Giovanni", 13, "Project Manager",
        employeeList, "COBOL Multiplatform"
    )
    giovanni.supervise()
    giovanni.showEmployees()

    employeeList.add(giovanni)

    val jetbrains = Manager("JetBrains", 14, "Manager", employeeList, "Kotlin")
    jetbrains.organize()
    jetbrains.showEmployees()
}
